File: py/indicators/john_ehlers/corona_spectrum/corona_spectrum.py
"""Ehlers' Corona Spectrum heatmap indicator."""

import math
import datetime

from ....entities.bar import Bar
from ....entities.bar_component import BarComponent, bar_component_value, DEFAULT_BAR_COMPONENT
from ....entities.quote import Quote
from ....entities.quote_component import QuoteComponent, quote_component_value, DEFAULT_QUOTE_COMPONENT
from ....entities.trade import Trade
from ....entities.trade_component import TradeComponent, trade_component_value, DEFAULT_TRADE_COMPONENT
from ....entities.scalar import Scalar
from ...core.identifier import Identifier
from ...core.metadata import Metadata
from ...core.build_metadata import build_metadata, OutputText
from ...core.component_triple_mnemonic import component_triple_mnemonic
from ...core.outputs.heatmap import Heatmap
from ..corona.corona import Corona
from ..corona.params import CoronaParams
from .params import Params


class CoronaSpectrum:
    """Ehlers' Corona Spectrum heatmap indicator.

    Outputs:
    - Value: per-bar heatmap column (decibels across filter bank)
    - DominantCycle: weighted-center-of-gravity dominant cycle estimate
    - DominantCycleMedian: 5-sample median of DominantCycle
    """

    def __init__(self, params: Params) -> None:
        invalid = "invalid corona spectrum parameters"

        cfg_min_raster = params.min_raster_value if params.min_raster_value != 0 else 6.0
        cfg_max_raster = params.max_raster_value if params.max_raster_value != 0 else 20.0
        cfg_min_param = params.min_parameter_value if params.min_parameter_value != 0 else 6.0
        cfg_max_param = params.max_parameter_value if params.max_parameter_value != 0 else 30.0
        cfg_hp_cutoff = params.high_pass_filter_cutoff if params.high_pass_filter_cutoff != 0 else 30

        if cfg_min_raster < 0:
            raise ValueError(f"{invalid}: MinRasterValue should be >= 0")
        if cfg_max_raster <= cfg_min_raster:
            raise ValueError(f"{invalid}: MaxRasterValue should be > MinRasterValue")

        min_param = math.ceil(cfg_min_param)
        max_param = math.floor(cfg_max_param)

        if min_param < 2:
            raise ValueError(f"{invalid}: MinParameterValue should be >= 2")
        if max_param <= min_param:
            raise ValueError(f"{invalid}: MaxParameterValue should be > MinParameterValue")
        if cfg_hp_cutoff < 2:
            raise ValueError(f"{invalid}: HighPassFilterCutoff should be >= 2")

        bc = params.bar_component if params.bar_component is not None else BarComponent.MEDIAN
        qc = params.quote_component if params.quote_component is not None else DEFAULT_QUOTE_COMPONENT
        tc = params.trade_component if params.trade_component is not None else DEFAULT_TRADE_COMPONENT

        self._bar_func = bar_component_value(bc)
        self._quote_func = quote_component_value(qc)
        self._trade_func = trade_component_value(tc)

        self._corona = Corona(CoronaParams(
            high_pass_filter_cutoff=cfg_hp_cutoff,
            minimal_period=int(min_param),
            maximal_period=int(max_param),
            decibels_lower_threshold=cfg_min_raster,
            decibels_upper_threshold=cfg_max_raster,
        ))

        self._min_parameter_value = float(min_param)
        self._max_parameter_value = float(max_param)
        self._parameter_resolution = float(self._corona.filter_bank_length - 1) / \
            (float(max_param) - float(min_param))
        self._min_raster_value = cfg_min_raster
        self._max_raster_value = cfg_max_raster

        comp_mn = component_triple_mnemonic(bc, qc, tc)
        self.mnemonic = f"cspect({cfg_min_raster:g}, {cfg_max_raster:g}, " \
            f"{min_param:g}, {max_param:g}, {cfg_hp_cutoff}{comp_mn})"
        self.description = "Corona spectrum " + self.mnemonic

        self._mnemonic_dc = f"cspect-dc({cfg_hp_cutoff}{comp_mn})"
        self._description_dc = "Corona spectrum dominant cycle " + self._mnemonic_dc
        self._mnemonic_dcm = f"cspect-dcm({cfg_hp_cutoff}{comp_mn})"
        self._description_dcm = "Corona spectrum dominant cycle median " + self._mnemonic_dcm

    def is_primed(self) -> bool:
        return self._corona.is_primed()

    def metadata(self) -> Metadata:
        return build_metadata(
            Identifier.CORONA_SPECTRUM,
            self.mnemonic,
            self.description,
            [
                OutputText(mnemonic=self.mnemonic, description=self.description),
                OutputText(mnemonic=self._mnemonic_dc, description=self._description_dc),
                OutputText(mnemonic=self._mnemonic_dcm, description=self._description_dcm),
            ],
        )

    def update(self, sample: float, t: datetime.datetime) -> tuple:
        """Returns (heatmap, dominant_cycle, dominant_cycle_median)."""
        if math.isnan(sample):
            return (Heatmap.empty(t, self._min_parameter_value,
                                  self._max_parameter_value, self._parameter_resolution),
                    float('nan'), float('nan'))

        primed = self._corona.update(sample)
        if not primed:
            return (Heatmap.empty(t, self._min_parameter_value,
                                  self._max_parameter_value, self._parameter_resolution),
                    float('nan'), float('nan'))

        bank = self._corona.filter_bank
        values = [0.0] * len(bank)
        value_min = math.inf
        value_max = -math.inf

        for i, f in enumerate(bank):
            v = f.decibels
            values[i] = v
            if v < value_min:
                value_min = v
            if v > value_max:
                value_max = v

        heatmap = Heatmap(t, self._min_parameter_value, self._max_parameter_value,
                          self._parameter_resolution, value_min, value_max, values)

        return heatmap, self._corona.dominant_cycle, self._corona.dominant_cycle_median

    def update_scalar(self, sample: Scalar) -> list:
        return self._update_entity(sample.time, sample.value)

    def update_bar(self, sample: Bar) -> list:
        return self._update_entity(sample.time, self._bar_func(sample))

    def update_quote(self, sample: Quote) -> list:
        return self._update_entity(sample.time, self._quote_func(sample))

    def update_trade(self, sample: Trade) -> list:
        return self._update_entity(sample.time, self._trade_func(sample))

    def _update_entity(self, t: datetime.datetime, sample: float) -> list:
        heatmap, dc, dcm = self.update(sample, t)
        return [heatmap, Scalar(t, dc), Scalar(t, dcm)]

File: rs/src/indicators/john_ehlers/corona_spectrum/corona_spectrum.rs
use std::any::Any;

use crate::entities::bar::Bar;
use crate::entities::bar_component::{component_value as bar_component_value, BarComponent};
use crate::entities::quote::Quote;
use crate::entities::quote_component::{
    component_value as quote_component_value, QuoteComponent, DEFAULT_QUOTE_COMPONENT,
};
use crate::entities::scalar::Scalar;
use crate::entities::trade::Trade;
use crate::entities::trade_component::{
    component_value as trade_component_value, TradeComponent, DEFAULT_TRADE_COMPONENT,
};
use crate::indicators::core::build_metadata::{build_metadata, OutputText};
use crate::indicators::core::component_triple_mnemonic::component_triple_mnemonic;
use crate::indicators::core::identifier::Identifier;
use crate::indicators::core::indicator::{Indicator, Output};
use crate::indicators::core::metadata::Metadata;
use crate::indicators::core::outputs::heatmap::Heatmap;
use crate::indicators::john_ehlers::corona::corona::{Corona, CoronaParams};

// ---------------------------------------------------------------------------
// Params
// ---------------------------------------------------------------------------

/// Parameters for the Corona Spectrum indicator.
pub struct CoronaSpectrumParams {
    /// Minimal raster value (z) of the heatmap, in decibels. Default 6.
    pub min_raster_value: f64,
    /// Maximal raster value (z) of the heatmap, in decibels. Default 20.
    pub max_raster_value: f64,
    /// Minimal ordinate (y) value — minimal cycle period. Default 6.
    pub min_parameter_value: f64,
    /// Maximal ordinate (y) value — maximal cycle period. Default 30.
    pub max_parameter_value: f64,
    /// High-pass filter cutoff (de-trending period). Default 30.
    pub high_pass_filter_cutoff: i32,
    /// Bar component. `None` → Median (hl/2).
    pub bar_component: Option<BarComponent>,
    /// Quote component. `None` → Mid.
    pub quote_component: Option<QuoteComponent>,
    /// Trade component. `None` → Price.
    pub trade_component: Option<TradeComponent>,
}

impl Default for CoronaSpectrumParams {
    fn default() -> Self {
        Self {
            min_raster_value: 6.0,
            max_raster_value: 20.0,
            min_parameter_value: 6.0,
            max_parameter_value: 30.0,
            high_pass_filter_cutoff: 30,
            bar_component: None,
            quote_component: None,
            trade_component: None,
        }
    }
}

// ---------------------------------------------------------------------------
// Output enum
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum CoronaSpectrumOutput {
    Value = 1,
    DominantCycle = 2,
    DominantCycleMedian = 3,
}

// ---------------------------------------------------------------------------
// Indicator
// ---------------------------------------------------------------------------

/// Ehlers' Corona Spectrum heatmap indicator.
///
/// Measures cyclic activity over a cycle period range in a bank of contiguous
/// bandpass filters. Outputs a heatmap column (decibels), dominant cycle, and
/// dominant cycle median.
pub struct CoronaSpectrum {
    mnemonic: String,
    description: String,
    mnemonic_dc: String,
    description_dc: String,
    mnemonic_dcm: String,
    description_dcm: String,
    corona: Corona,
    min_parameter_value: f64,
    max_parameter_value: f64,
    parameter_resolution: f64,
    bar_func: fn(&Bar) -> f64,
    quote_func: fn(&Quote) -> f64,
    trade_func: fn(&Trade) -> f64,
}

impl CoronaSpectrum {
    pub fn new(p: &CoronaSpectrumParams) -> Result<Self, String> {
        let invalid = "invalid corona spectrum parameters";

        let min_raster = if p.min_raster_value == 0.0 { 6.0 } else { p.min_raster_value };
        let max_raster = if p.max_raster_value == 0.0 { 20.0 } else { p.max_raster_value };
        let min_pv = if p.min_parameter_value == 0.0 { 6.0 } else { p.min_parameter_value };
        let max_pv = if p.max_parameter_value == 0.0 { 30.0 } else { p.max_parameter_value };
        let hp = if p.high_pass_filter_cutoff == 0 { 30 } else { p.high_pass_filter_cutoff };

        if min_raster < 0.0 {
            return Err(format!("{}: MinRasterValue should be >= 0", invalid));
        }
        if max_raster <= min_raster {
            return Err(format!("{}: MaxRasterValue should be > MinRasterValue", invalid));
        }

        let min_param = min_pv.ceil();
        let max_param = max_pv.floor();

        if min_param < 2.0 {
            return Err(format!("{}: MinParameterValue should be >= 2", invalid));
        }
        if max_param <= min_param {
            return Err(format!("{}: MaxParameterValue should be > MinParameterValue", invalid));
        }
        if hp < 2 {
            return Err(format!("{}: HighPassFilterCutoff should be >= 2", invalid));
        }

        let bc = p.bar_component.unwrap_or(BarComponent::Median);
        let qc = p.quote_component.unwrap_or(DEFAULT_QUOTE_COMPONENT);
        let tc = p.trade_component.unwrap_or(DEFAULT_TRADE_COMPONENT);

        let corona = Corona::new(&CoronaParams {
            high_pass_filter_cutoff: hp,
            minimal_period: min_param as i32,
            maximal_period: max_param as i32,
            decibels_lower_threshold: min_raster,
            decibels_upper_threshold: max_raster,
        })?;

        let comp_mn = component_triple_mnemonic(bc, qc, tc);
        let parameter_resolution =
            (corona.filter_bank_length() as f64 - 1.0) / (max_param - min_param);

        let mnemonic = format!(
            "cspect({}, {}, {}, {}, {}{})",
            min_raster, max_raster, min_param, max_param, hp, comp_mn
        );
        let mnemonic_dc = format!("cspect-dc({}{})", hp, comp_mn);
        let mnemonic_dcm = format!("cspect-dcm({}{})", hp, comp_mn);

        Ok(Self {
            description: format!("Corona spectrum {}", mnemonic),
            mnemonic,
            description_dc: format!("Corona spectrum dominant cycle {}", mnemonic_dc),
            mnemonic_dc,
            description_dcm: format!("Corona spectrum dominant cycle median {}", mnemonic_dcm),
            mnemonic_dcm,
            corona,
            min_parameter_value: min_param,
            max_parameter_value: max_param,
            parameter_resolution,
            bar_func: bar_component_value(bc),
            quote_func: quote_component_value(qc),
            trade_func: trade_component_value(tc),
        })
    }

    /// Feed the next sample and return (heatmap, dominant_cycle, dominant_cycle_median).
    pub fn update(&mut self, sample: f64, time: i64) -> (Heatmap, f64, f64) {
        if sample.is_nan() {
            return (
                Heatmap::empty(time, self.min_parameter_value, self.max_parameter_value, self.parameter_resolution),
                f64::NAN,
                f64::NAN,
            );
        }

        let primed = self.corona.update(sample);
        if !primed {
            return (
                Heatmap::empty(time, self.min_parameter_value, self.max_parameter_value, self.parameter_resolution),
                f64::NAN,
                f64::NAN,
            );
        }

        let bank = self.corona.filter_bank();
        let mut values = Vec::with_capacity(bank.len());
        let mut value_min = f64::INFINITY;
        let mut value_max = f64::NEG_INFINITY;

        for f in bank {
            let v = f.decibels;
            values.push(v);
            if v < value_min { value_min = v; }
            if v > value_max { value_max = v; }
        }

        let heatmap = Heatmap::new(
            time,
            self.min_parameter_value,
            self.max_parameter_value,
            self.parameter_resolution,
            value_min,
            value_max,
            values,
        );

        (heatmap, self.corona.dominant_cycle(), self.corona.dominant_cycle_median())
    }
}

impl Indicator for CoronaSpectrum {
    fn is_primed(&self) -> bool {
        self.corona.is_primed()
    }

    fn metadata(&self) -> Metadata {
        build_metadata(
            Identifier::CoronaSpectrum,
            &self.mnemonic,
            &self.description,
            &[
                OutputText { mnemonic: self.mnemonic.clone(), description: self.description.clone() },
                OutputText { mnemonic: self.mnemonic_dc.clone(), description: self.description_dc.clone() },
                OutputText { mnemonic: self.mnemonic_dcm.clone(), description: self.description_dcm.clone() },
            ],
        )
    }

    fn update_bar(&mut self, bar: &Bar) -> Output {
        let sample = (self.bar_func)(bar);
        let (h, dc, dcm) = self.update(sample, bar.time);
        vec![
            Box::new(h) as Box<dyn Any>,
            Box::new(Scalar::new(bar.time, dc)),
            Box::new(Scalar::new(bar.time, dcm)),
        ]
    }

    fn update_quote(&mut self, quote: &Quote) -> Output {
        let sample = (self.quote_func)(quote);
        let (h, dc, dcm) = self.update(sample, quote.time);
        vec![
            Box::new(h) as Box<dyn Any>,
            Box::new(Scalar::new(quote.time, dc)),
            Box::new(Scalar::new(quote.time, dcm)),
        ]
    }

    fn update_trade(&mut self, trade: &Trade) -> Output {
        let sample = (self.trade_func)(trade);
        let (h, dc, dcm) = self.update(sample, trade.time);
        vec![
            Box::new(h) as Box<dyn Any>,
            Box::new(Scalar::new(trade.time, dc)),
            Box::new(Scalar::new(trade.time, dcm)),
        ]
    }

    fn update_scalar(&mut self, scalar: &Scalar) -> Output {
        let (h, dc, dcm) = self.update(scalar.value, scalar.time);
        vec![
            Box::new(h) as Box<dyn Any>,
            Box::new(Scalar::new(scalar.time, dc)),
            Box::new(Scalar::new(scalar.time, dcm)),
        ]
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::testdata::testdata;
    const TOLERANCE: f64 = 1e-4;

    #[test]
    fn test_corona_spectrum_update() {
        let input = testdata::talib_input();

        struct Snap { i: usize, dc: f64, dcm: f64 }
        let snapshots = [
            Snap { i: 11,  dc: 17.7604672565, dcm: 17.7604672565 },
            Snap { i: 12,  dc: 6.0000000000,  dcm: 6.0000000000 },
            Snap { i: 50,  dc: 15.9989078712, dcm: 15.9989078712 },
            Snap { i: 100, dc: 14.7455497547, dcm: 14.7455497547 },
            Snap { i: 150, dc: 17.5000000000, dcm: 17.2826036069 },
            Snap { i: 200, dc: 19.7557338512, dcm: 20.0000000000 },
            Snap { i: 251, dc: 6.0000000000,  dcm: 6.0000000000 },
        ];

        let mut x = CoronaSpectrum::new(&CoronaSpectrumParams::default()).unwrap();

        let mut si = 0;
        for (i, &v) in input.iter().enumerate() {
            let (h, dc, dcm) = x.update(v, i as i64);

            // Heatmap axis invariants.
            assert_eq!(h.parameter_first, 6.0, "[{}] parameter_first", i);
            assert_eq!(h.parameter_last, 30.0, "[{}] parameter_last", i);
            assert_eq!(h.parameter_resolution, 2.0, "[{}] parameter_resolution", i);

            if !x.is_primed() {
                assert!(h.is_empty(), "[{}] expected empty heatmap before priming", i);
                assert!(dc.is_nan(), "[{}] expected NaN dc before priming", i);
                assert!(dcm.is_nan(), "[{}] expected NaN dcm before priming", i);
                continue;
            }

            assert_eq!(h.values.len(), 49, "[{}] heatmap values length", i);

            if si < snapshots.len() && snapshots[si].i == i {
                assert!(
                    (snapshots[si].dc - dc).abs() < TOLERANCE,
                    "[{}] dc: expected {}, got {}", i, snapshots[si].dc, dc
                );
                assert!(
                    (snapshots[si].dcm - dcm).abs() < TOLERANCE,
                    "[{}] dcm: expected {}, got {}", i, snapshots[si].dcm, dcm
                );
                si += 1;
            }
        }

        assert_eq!(si, snapshots.len(), "did not hit all snapshots");
    }

    #[test]
    fn test_corona_spectrum_primes_at_bar_11() {
        let mut x = CoronaSpectrum::new(&CoronaSpectrumParams::default()).unwrap();
        assert!(!x.is_primed());

        let input = testdata::talib_input();
        let mut primed_at: Option<usize> = None;

        for (i, &v) in input.iter().enumerate() {
            x.update(v, i as i64);
            if x.is_primed() && primed_at.is_none() {
                primed_at = Some(i);
            }
        }

        assert_eq!(primed_at, Some(11), "expected priming at index 11");
    }

    #[test]
    fn test_corona_spectrum_nan_input() {
        let mut x = CoronaSpectrum::new(&CoronaSpectrumParams::default()).unwrap();
        let (h, dc, dcm) = x.update(f64::NAN, 0);
        assert!(h.is_empty());
        assert!(dc.is_nan());
        assert!(dcm.is_nan());
        assert!(!x.is_primed());
    }

    #[test]
    fn test_corona_spectrum_metadata() {
        let x = CoronaSpectrum::new(&CoronaSpectrumParams::default()).unwrap();
        let md = x.metadata();

        assert_eq!(md.identifier, Identifier::CoronaSpectrum);
        assert_eq!(md.mnemonic, "cspect(6, 20, 6, 30, 30, hl/2)");
        assert_eq!(md.description, "Corona spectrum cspect(6, 20, 6, 30, 30, hl/2)");
        assert_eq!(md.outputs.len(), 3);

        assert_eq!(md.outputs[0].kind, 1); // Value
        assert_eq!(md.outputs[0].mnemonic, "cspect(6, 20, 6, 30, 30, hl/2)");
        assert_eq!(md.outputs[1].kind, 2); // DominantCycle
        assert_eq!(md.outputs[1].mnemonic, "cspect-dc(30, hl/2)");
        assert_eq!(md.outputs[2].kind, 3); // DominantCycleMedian
        assert_eq!(md.outputs[2].mnemonic, "cspect-dcm(30, hl/2)");
    }

    #[test]
    fn test_corona_spectrum_update_bar() {
        let input = testdata::talib_input();
        let mut x = CoronaSpectrum::new(&CoronaSpectrumParams::default()).unwrap();

        // Prime with 50 samples.
        for (i, &v) in input.iter().take(50).enumerate() {
            x.update(v, i as i64);
        }

        let bar = Bar::new(100, 100.0, 100.0, 100.0, 100.0, 0.0);
        let out = x.update_bar(&bar);
        assert_eq!(out.len(), 3);
    }

    #[test]
    fn test_corona_spectrum_invalid_params() {
        // MaxRasterValue <= MinRasterValue
        assert!(CoronaSpectrum::new(&CoronaSpectrumParams {
            min_raster_value: 10.0,
            max_raster_value: 10.0,
            ..CoronaSpectrumParams::default()
        }).is_err());

        // MinParameterValue < 2
        assert!(CoronaSpectrum::new(&CoronaSpectrumParams {
            min_parameter_value: 1.0,
            ..CoronaSpectrumParams::default()
        }).is_err());

        // MaxParameterValue <= MinParameterValue
        assert!(CoronaSpectrum::new(&CoronaSpectrumParams {
            min_parameter_value: 20.0,
            max_parameter_value: 20.0,
            ..CoronaSpectrumParams::default()
        }).is_err());

        // HighPassFilterCutoff < 2
        assert!(CoronaSpectrum::new(&CoronaSpectrumParams {
            high_pass_filter_cutoff: 1,
            ..CoronaSpectrumParams::default()
        }).is_err());
    }

    #[test]
    fn test_corona_spectrum_custom_ranges() {
        let x = CoronaSpectrum::new(&CoronaSpectrumParams {
            min_raster_value: 4.0,
            max_raster_value: 25.0,
            min_parameter_value: 8.7,  // ceils to 9
            max_parameter_value: 40.4, // floors to 40
            high_pass_filter_cutoff: 20,
            ..CoronaSpectrumParams::default()
        }).unwrap();

        assert_eq!(x.min_parameter_value, 9.0);
        assert_eq!(x.max_parameter_value, 40.0);
        assert_eq!(x.mnemonic, "cspect(4, 25, 9, 40, 20, hl/2)");
    }
}
